sdk/trace/span.go

```diff
@@ -575,81 +577,65 @@ func (s *span) addChild() {
 	s.mu.Unlock()
 }
 
-func (*span) private() {}
+func (*recordingSpan) private() {}
 
-func startSpanInternal(ctx context.Context, tr *tracer, name string, o *trace.SpanConfig) *span {
-	span := &span{}
+// runtimeTrace starts a "runtime/trace".Task for the span and returns a
+// context containing the task.
+func (s *recordingSpan) runtimeTrace(ctx context.Context) context.Context {
+	if !rt.IsEnabled() {
+		// Avoid additional overhead if runtime/trace is not enabled.
+		return ctx
+	}
+	nctx, task := rt.NewTask(ctx, s.name)
 
-	provider := tr.provider
+	s.mu.Lock()
+	s.executionTracerTaskEnd = task.End
+	s.mu.Unlock()
 
-	// If told explicitly to make this a new root use a zero value SpanContext
-	// as a parent which contains an invalid trace ID and is not remote.
-	var psc trace.SpanContext
-	if o.NewRoot() {
-		ctx = trace.ContextWithSpanContext(ctx, psc)
-	} else {
-		psc = trace.SpanContextFromContext(ctx)
-	}
+	return nctx
+}
 
-	// If there is a valid parent trace ID, use it to ensure the continuity of
-	// the trace. Always generate a new span ID so other components can rely
-	// on a unique span ID, even if the Span is non-recording.
-	var tid trace.TraceID
-	var sid trace.SpanID
-	if !psc.TraceID().IsValid() {
-		tid, sid = provider.idGenerator.NewIDs(ctx)
-	} else {
-		tid = psc.TraceID()
-		sid = provider.idGenerator.NewSpanID(ctx, tid)
-	}
+// nonRecordingSpan is a minimal implementation of the OpenTelemetry Span API
+// that wraps a SpanContext. It performs no operations other than to return
+// the wrapped SpanContext or TracerProvider that created it.
+type nonRecordingSpan struct {
+	// tracer is the SDK tracer that created this span.
+	tracer *tracer
```

Comment:
Could you embed a `noopSpan`, then only implement `SpanContext()`?

Reply:
`noopSpan` is in the API and unexported, so it would be unavailable to use here.

Reply:
oh, right! this is still possible, though maybe not worth it, using `SpanFromContext(context.Background())` to get one of the `noopSpan` objects as a `trace.Span`. looks like this:

https://play.golang.org/p/lBS2v7L1TWg

Reply:
I think ["a little copying is better than a little dependency"](https://go-proverbs.github.io/) applies here. It is probably a good exercise to take a fresh look at the span types and see if we could restructure them to be universal to the project, but it seems that would be a good independent change for its own PR.